File: python3/Linked Lists/linked_list_reversal.py
from ds import ListNode

"""
Definition of ListNode:
class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


def linked_list_reversal(head: ListNode) -> ListNode:
    curr_node, prev_node = head, None
    # Reverse the direction of each node's pointer until 'curr_node' 
    # is null.
    while curr_node:
        next_node = curr_node.next
        curr_node.next = prev_node
        prev_node = curr_node
        curr_node = next_node
    # 'prev_node' will be pointing at the head of the reversed linked 
    # list.
    return prev_node

File: cpp/Linked Lists/linked_list_reversal.cpp
/**
 * Definition of ListNode:
 * struct ListNode {
 *     int val;
 *     ListNode* next;
 *     ListNode(int val = 0, ListNode* next = nullptr) : val(val), next(next) {}
 * };
 */

ListNode* linkedListReversal(ListNode* head) {
    ListNode* currNode = head;      
    ListNode* prevNode = nullptr;   
    // Reverse the direction of each node's pointer until 'currNode' 
    // is null.
    while (currNode != nullptr) {
        ListNode* nextNode = currNode->next;  
        currNode->next = prevNode;            
        prevNode = currNode;                 
        currNode = nextNode;                  
    }
    // 'prevNode' will be pointing at the head of the reversed linked 
    // list.
    return prevNode;
}